docs: add hooks specification to the sidebar (#174)
slackapi/slack-cli: docs/reference/hooks/index.md

| config.watch                          | Object with configuration settings for file-watching.                                                                                                                                                                                                                                                                                                                                                           | Optional |
| config.sdk-managed-connection-enabled | Boolean specifying whether the WebSocket connection between the CLI and Slack should be managed by the CLI or by the SDK during `slack run` executions. If `true`, the SDK will manage this connection. If `false` or not provided, the CLI will manage this connection.                                                                                                                                        | Optional |

> | config.trigger-paths                  | Array of strings that are paths to files of trigger definitions.                                                                                                                                                                                                                                                                                                                                                | Optional |

Author: 📚 note: This reference was added in recent commit https://github.com/slackapi/slack-cli/pull/174/commits/80689d9568625a5fc6192290fc1c8e8573aec7b1 but I think separate reference for the `config` options might be nice to follow up with.

👁️‍🗨️ note: Right now this option is also documented at: https://docs.slack.dev/tools/deno-slack-sdk/guides/using-triggers/#custom-path

# Hooks

> Updated: May 2024

Author: @lukegalbraithrussell I was unsure about this - thank you for commenting! 👾 ✨ 

In quick thoughts I agree that we can remove this, but I'm curious what others think as well? 🔍 @mwbrooks @vegeris @WilliamBergamin

Removing the "last updated" date seems alright IMO since:

- Changes are in public version control
- Troubleshooting steps suggest using the latest versions
- Updates to this document will be shared elsewhere

> Updated: May 2024

Reviewer: further elaboration for the curious - putting "last updated" just makes the docs seem out of date when they might not be. Also it just ages the doc

(flashbacks to when i found a doc last year that said "stay tuned for this in 2019!" 💀

> Updated: May 2024

Reviewer 2: I agree that we can remove the date, if folks feel strongly about it.

FWIW git provides a versioning history of the specification. Until we find a need for the explicit semver version number, we can at least see revisions to the specification on github.com.

## Background {#background}

Communication between the CLI and the application SDK is managed by a project-level configuration file called `hooks.json`. This file is included in our app templates and defines script _hooks_.

Hooks are small scripts that are _executed_ by the CLI and _implemented_ by the SDK. These commands perform actions on the project.

The `hooks.json` file allows the CLI and SDK a standard way to communicate while remaining decoupled and abstracted. This interface is a key design of the Slack CLI: many application and project level tasks are delegated from the CLI to the SDK. This delegation, decoupling and abstraction allows for language-agnostic SDK implementations.

> ## Core Concepts {#core-concepts}

Reviewer: ```suggestion
## Core concepts {#core-concepts}
```

For example, the hook name `run-v2` may be the successor to the hook named `run`. The SDK can implement either hook and the CLI will trigger the latest version, possibly falling back to earlier versions of the hook where applicable. The CLI can also warn of impending removal of older hooks, providing hints to the developer when tooling behaviour changes.

## Hook Specification {#specification}

Hooks are entry points for the CLI to initiate inter-process communication with the SDK. SDKs should implement one hook: `get-hooks`. It is the recommended approach for SDKs to enable communication with the CLI (for more details, see the [Hook Resolution](#hook-resolution) section).

A hook encapsulates an isolated piece of functionality that the SDK provides to app developers; the CLI delegates execution of the functionality to the SDK via hook invocation in a separate process. SDKs are responsible for ensuring that hook processes exit with a status code of 0; otherwise the CLI will surface the `sdk_hook_invocation_failed` error to the end-user and report the hook process’ `STDOUT` and `STDERR` to the CLI process’ `STDOUT`.

> ## List of hooks {#list-of-hooks}

Reviewer: ```suggestion
## Hooks list {#hooks-list}
```

The following Slack CLI commands invoke hooks:

| Command               | Hook(s)                           |
| --------------------- | --------------------------------- |

> | `slack manifest info` | `get-manifest`                    |
> | `slack deploy`        | `get-manifest`, `build`, `deploy` |
> | `slack run`           | `get-manifest`, `start`           |
> | `slack update`        | `check-update`                    |

Reviewer: ```suggestion
| `slack manifest info` | `get-manifest`                    |
| `slack deploy`        | [`get-manifest`](#get-manifest), [`build`](#build), ['deploy'](#deploy) |
| `slack run`              | ['get-manifest'](#get-manifest), [`start`](#start)           |
| `slack update`        | [`check-update`](#check-update)             |
```

- Each custom function's source code should be bundled into a single `.js` file with its `callback_id` as its filename, and exist inside a `functions/` subdirectory within the app bundle. For example, a custom function with a callback_id of `greeting_function` should bundle all of its code, including dependencies, into a `greeting_function.js` file that would exist at `functions/greeting_function.js` within the application bundle.

The above requirements come from the [deno-slack-runtime](https://github.com/slackapi/deno-slack-runtime) project, which implements the expected Slack deployment bundle format. It contains a hard-coded [reference](https://github.com/slackapi/deno-slack-runtime/blob/main/src/mod.ts#L73) to the above-mentioned `functions/` sub-directory, and combines it with the [specific custom function `callback_id`](https://github.com/slackapi/deno-slack-runtime/blob/main/src/mod.ts#L17-L19) to resolve an import path for userland function source code.

> _Note: This hook should only be implemented by official Slack SDKs and is only relevant to apps [deployed to Slack's managed infrastructure](https://tools.slack.dev/deno-slack-sdk/guides/deploying-to-slack)._

Reviewer: ```suggestion
:::warning[This hook should only be implemented by official Slack SDKs and is only relevant to apps [deployed to Slack's managed infrastructure](https://tools.slack.dev/deno-slack-sdk/guides/deploying-to-slack).
:::
```

No further output over STDOUT required from the SDK other than writing the application bundle to the filesystem location specified by the `--output` flag.

#### Support

| Deno | Bolt JS | Bolt Python | Bolt Java |
| ---- | ------- | ----------- | --------- |

> | ✅   | ❌      | ❌          | ❌        |

Reviewer: ```suggestion
| ✅ Supported  | ❌ Unsupported     | ❌ Unsupported         | ❌ Unsupported        |
```

##### Input

Several parameters are passed to the SDK over `STDIN` as JSON. The format of this JSON payload is as follows:

```json
{
  "body": {
    "type": "app_deleted",
    ...
  },
  "context": {
    "bot_access_token": "xoxb-1234",
    "app_id": "A1234",
    "team_id": "T1234",
    "variables": {}
  }
}
```

| Field                    | Description                                                                                                                                                                                                                             | Required |
| ------------------------ | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | -------- |

> | body                     | Object whose keys represent the incoming Slack event payload as described in [Events API Event Types](https://docs.slack.dev/reference/events?APIs=Events). The particular content of this key is dependent on the incoming event type. | Yes      |
> | context                  | Object representing variables relevant for the locally-running application.                                                                                                                                                             | Yes      |
> | context.bot_access_token | String; a bot access token that the SDK can provide to developer functions for issuing calls to the Slack API.                                                                                                                          | Yes      |
> | context.app_id           | String; the current application ID.                                                                                                                                                                                                     | Yes      |
> | context.team_id          | String; the ID of the team or workspace the locally-running app is installed to.                                                                                                                                                        | Yes      |
> | context.variables        | Object containing environment variables that may or may not be defined by the application via a `.env` file in the root of the application directory.                                                                                   | Yes      |

Reviewer: ```suggestion
| body                     | Object whose keys represent the incoming Slack event payload as described in [Events API Event Types](https://docs.slack.dev/reference/events?APIs=Events). The particular content of this key is dependent on the incoming event type. | Required      |
| context                  | Object representing variables relevant for the locally-running application.                                                                                                                                                             | Required      |
| context.bot_access_token | String; a bot access token that the SDK can provide to developer functions for issuing calls to the Slack API.                                                                                                                          | Required      |
| context.app_id           | String; the current application ID.                                                                                                                                                                                                     | Required      |
| context.team_id          | String; the ID of the team or workspace the locally-running app is installed to.                                                                                                                                                        | Required      |
| context.variables        | Object containing environment variables that may or may not be defined by the application via a `.env` file in the root of the application directory.                                                                                   | Required      |
```

screen reader accessibility

Each incoming event from the socket connection will invoke this hook separately. As such, this hook's response to `STDOUT` should be the JSON response to the Slack event sent to the app. The CLI will handle [acknowledging events](https://docs.slack.dev/apis/events-api/using-socket-mode/#acknowledge) by sending back the proper `envelope_id` attribute to the Slack backend. Therefore, the SDK’s `start` hook response `STDOUT` should be the `payload` of the response and nothing else. It is recommended to use the `v2` (`message-boundaries`) [protocol](#protocol) to more easily delineate logging/diagnostics from event responses to be sent to Slack.

##### Support

| Deno | Bolt JS | Bolt Python | Bolt Java |
| ---- | ------- | ----------- | --------- |

> | ✅   | ❌      | ❌          | ❌        |

Reviewer: ```suggestion
| ✅ Supported  | ❌ Unsupported      | ❌ Unsupported         | ❌ Unsupported        |
```

Any `STDOUT` received from the this hook would be immediately streamed to the CLI process’ `STDOUT`.

##### Support

| Deno | Bolt JS | Bolt Python | Bolt Java |
| ---- | ------- | ----------- | --------- |

> | ❌   | ✅      | ✅          | ❌        |

Reviewer: ```suggestion
| ❌ Unsupported  | ✅ Supported     | ✅ Supported         | ❌ Unsupported       |
```

| releases.error    | Object with a single key, "message", which has a string value containing error information.                                                                                               | No       |
| message           | String containing any additional details that should be surfaced to the user. For example, this could include how to manually update or a warning that certain files will be overwritten. | No       |
| url               | String containing a URL where one can learn more about the release.                                                                                                                       | No       |
| error             | Object with a single key, "message", which has a string value containing error information.                                                                                               | No       |

#### Support

| Deno | Bolt JS | Bolt Python | Bolt Java |
| ---- | ------- | ----------- | --------- |

> | ✅   | ✅      | ❌          | ❌        |

Reviewer: ```suggestion
| ✅ Supported  | ✅ Supported     | ❌ Unsupported          | ❌ Unsupported       |
```

### `install-update` (optional) {#install-update}

Update the SDK version and/or any dependencies required by the app.

#### Output

The format for the output JSON is as follows:

```json
{
  "name": "",
  "releases": [
    {
      "name": "",
      "current": "",
      "latest": "",
      "update": true,
      "breaking": false,
      "message": "",
      "url": "",
      "error": ""
    },
    ...
  ],
  "message": "",
  "url": "",
  "error": ""
}
```

| Field             | Description                                                                                                                                                                               | Required |
| ----------------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | -------- |

> | name              | String containing the name corresponding to the overall package/library, in which individual component releases are bundled.                                                              | Yes      |
> | releases          | Array of objects representing individual releases.                                                                                                                                        | Yes      |
> | releases.name     | String containing the name of the dependency.                                                                                                                                             | Yes      |
> | releases.current  | String containing the current version.                                                                                                                                                    | No       |
> | releases.latest   | String containing the latest version available.                                                                                                                                           | No       |
> | releases.update   | Boolean indicating whether there is an update available.                                                                                                                                  | No       |
> | releases.breaking | Boolean indicating whether the update is breaking.                                                                                                                                        | No       |
> | releases.message  | String containing a message about the dependency.                                                                                                                                         | No       |
> | releases.url      | String containing a URL with update information.                                                                                                                                          | No       |
> | releases.error    | Object with a single key, "message", which has a string value containing error information.                                                                                               | No       |
> | message           | String containing any additional details that should be surfaced to the user. For example, this could include how to manually update or a warning that certain files will be overwritten. | No       |
> | url               | String containing a URL where one can learn more about the release.                                                                                                                       | No       |
> | error             | Object with a single key, "message", which has a string value containing error information.                                                                                               | No       |

Reviewer: ```suggestion
| name              | String containing the name corresponding to the overall package/library, in which individual component releases are bundled.                                                              | Required      |
| releases          | Array of objects representing individual releases.                                                                                                                                        | Required      |
| releases.name     | String containing the name of the dependency.                                                                                                                                             | Required      |
| releases.current  | String containing the current version.                                                                                                                                                    | Optional       |
| releases.latest   | String containing the latest version available.                                                                                                                                           | Optional       |
| releases.update   | Boolean indicating whether there is an update available.                                                                                                                                  | Optional       |
| releases.breaking | Boolean indicating whether the update is breaking.                                                                                                                                        | Optional       |
| releases.message  | String containing a message about the dependency.                                                                                                                                         | Optional       |
| releases.url      | String containing a URL with update information.                                                                                                                                          | Optional       |
| releases.error    | Object with a single key, "message", which has a string value containing error information.                                                                                               | Optional       |
| message           | String containing any additional details that should be surfaced to the user. For example, this could include how to manually update or a warning that certain files will be overwritten. | No       |
| url               | String containing a URL where one can learn more about the release.                                                                                                                       | Optional       |
| error             | Object with a single key, "message", which has a string value containing error information.                                                                                               | Optional       |
```

#### Support

> | Deno | Bolt JS | Bolt Python | Bolt Java |

Reviewer: oh i just noticed the table names, I kinda think full names would be better but idk how that would look on the page - would maybe be too long?
```suggestion
| Deno Slack SDK | Bolt for JS | Bolt for Python | Bolt for Java |
```

Author: @lukegalbraithrussell What a great call! I opted for an even more worded "Bolt for JavaScript" in ba38d37.

| ---- | ------- | ----------- | --------- |

> | ✅   | ✅      | ❌          | ❌        |

Reviewer: ```suggestion
| ✅  Supported  | ✅ Supported     | ❌ Unsupported         | ❌ Unsupported       |
```

### `get-trigger` (optional) {#get-trigger}

Used by the `triggers create` and `triggers update` CLI commands when the `--trigger-def` argument is passed. This hook should convert the app developer's code into a valid JSON blob that the CLI can upload to the trigger endpoints. The CLI handles adding the `trigger_id` when using the `trigger update` command.

#### Output

Below is a sample JSON blob representing the trigger:

```json
{
  "type": "shortcut",
  "name": "Submit an issue",
  "description": "Submit an issue to the channel",
  "workflow": "#/workflows/submit_issue",
  "workflow_app_id": "A0168GS8ZFV",
  "inputs": {
    "channel": {
      "value": "{{data.channel_id}}"
    },
    "interactivity": {
      "value": "{{data.interactivity}}"
    }
  }
}
```

#### Support

| Deno | Bolt JS | Bolt Python | Bolt Java |
| ---- | ------- | ----------- | --------- |

> | ✅   | ✅      | ❌          | ❌        |

Reviewer: ```suggestion
| ✅ Supported  | ✅ Supported      | ❌ Unsupported          | ❌ Unsupported       |
```

#### Output

The format for the output JSON is as follows:

```json
{
  "versions": [
    {
      "name": "",
      "current": "",
      "message": "",
      "error": ""
    },
    ...
  ]
}
```

| Field            | Description                                               | Required |
| ---------------- | --------------------------------------------------------- | -------- |

> | versions         | Array of objects containing runtime details.              | Yes      |
> | versions.name    | String containing the name of the runtime dependency.     | Yes      |
> | versions.current | String containing the current system version.             | Yes      |
> | versions.message | String containing a message about the runtime dependency. | No       |
> | versions.error   | String containing error information for the runtime.      | No       |

Reviewer: ```suggestion
| versions         | Array of objects containing runtime details.              | Required      |
| versions.name    | String containing the name of the runtime dependency.     | Required      |
| versions.current | String containing the current system version.             | Required      |
| versions.message | String containing a message about the runtime dependency. | Optional       |
| versions.error   | String containing error information for the runtime.      | Optional       |
```

#### Support

| Deno | Bolt JS | Bolt Python | Bolt Java |
| ---- | ------- | ----------- | --------- |

> | ❌   | ✅      | ✅          | ❌        |

Reviewer: ```suggestion
| ❌ Unsupported  | ✅ Supported     | ✅ Supported         | ❌ Unsupported       |
```

    "install-update": "...",
    "get-trigger": "...",
    "doctor": "...",
    "deploy": "..."
  },
  "config": {
    "protocol-version": ["message-boundaries"],
    "sdk-managed-connection-enabled": false
  }
}
```

| Field                                 | Description                                                                                                                                                                                                                                                                                                                                                                                                     | Required |
| ------------------------------------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | -------- |

> | runtime                               | String denoting the target runtime for app functions to execute in. For apps deployed to Slack's managed infrastructure, the only accepted value today is `deno`.                                                                                                                                                                                                                                               | Yes      |
> | hooks                                 | Object whose keys must match the hook names outlined in the above [Hooks Specification](#specification). Arguments can be provided within this string by separating them with spaces.                                                                                                                                                                                                                           | Yes      |
> | config                                | Object of key-value settings.                                                                                                                                                                                                                                                                                                                                                                                   | No       |
> | config.protocol-version               | Array of strings representing the named CLI-SDK protocols supported by the SDK, in descending order of support, as in the first element in the array defines the preferred protocol for use by the SDK, the second element defines the next-preferred protocol, and so on. The only supported named protocol currently is `message-boundaries`. The CLI will use the v1 protocol if this field is not provided. | No       |
> | config.watch                          | String with configuration settings for file-watching.                                                                                                                                                                                                                                                                                                                                                           | No       |
> | config.sdk-managed-connection-enabled | Boolean specifying whether the WebSocket connection between the CLI and Slack should be managed by the CLI or by the SDK during `slack run` executions. If `true`, the SDK will manage this connection. If `false` or not provided, the CLI will manage this connection.                                                                                                                                        | No       |

Reviewer: ```suggestion
| runtime                               | String denoting the target runtime for app functions to execute in. For apps deployed to Slack's managed infrastructure, the only accepted value today is `deno`.                                                                                                                                                                                                                                               | Required      |
| hooks                                 | Object whose keys must match the hook names outlined in the above [Hooks Specification](#specification). Arguments can be provided within this string by separating them with spaces.                                                                                                                                                                                                                           | Required      |
| config                                | Object of key-value settings.                                                                                                                                                                                                                                                                                                                                                                                   | Optional       |
| config.protocol-version               | Array of strings representing the named CLI-SDK protocols supported by the SDK, in descending order of support, as in the first element in the array defines the preferred protocol for use by the SDK, the second element defines the next-preferred protocol, and so on. The only supported named protocol currently is `message-boundaries`. The CLI will use the v1 protocol if this field is not provided. | No       |
| config.watch                          | String with configuration settings for file-watching.                                                                                                                                                                                                                                                                                                                                                           | Optional       |
| config.sdk-managed-connection-enabled | Boolean specifying whether the WebSocket connection between the CLI and Slack should be managed by the CLI or by the SDK during `slack run` executions. If `true`, the SDK will manage this connection. If `false` or not provided, the CLI will manage this connection.                                                                                                                                        | Optional       |
```
